examples/common/mod.rs

```diff
@@ -36,34 +43,53 @@ pub(crate) fn qos_from_str(s: &str) -> Result<mqtt3::proto::QoS, String> {
 
 pub(crate) const PACKET_STATS_WINDOW_SIZE: usize = 60;
 
+type History = heapless::HistoryBuffer<u128, PACKET_STATS_WINDOW_SIZE>;
+
 #[allow(dead_code)]
 pub(crate) struct PacketStats {
     start_time: std::time::Instant,
     current: u128,
-    history: heapless::HistoryBuffer<u128, PACKET_STATS_WINDOW_SIZE>,
+    bytes: u128,
+    message_history: History,
+    byte_history: History,
 }
 
 impl PacketStats {
     #[allow(dead_code)]
-    pub(crate) fn count(&mut self, num: usize) {
+    pub(crate) fn log_metrics(&mut self, num: usize, bytes: usize) {
         self.current += num as u128;
+        self.bytes += bytes as u128;
 
         let now = std::time::Instant::now();
         let elapsed = now.duration_since(self.start_time);
         if elapsed > std::time::Duration::from_secs(1) {
-            self.history.write(self.current);
+            // Message metrics
+            self.message_history.write(self.current);
 
-            log::info!(
-                "{:>8} | 1s average: {:>8}/s) | {}s average: {:>8}/s{}",
+            log_metric(
+                "messages processed",
                 self.current,
-                self.current * 1_000_000 / elapsed.as_micros(),
-                PACKET_STATS_WINDOW_SIZE,
-                self.history.as_slice().iter().sum::<u128>() * 1_000_000 / (self.history.len() as u128) / elapsed.as_micros(),
-                if self.history.len() < PACKET_STATS_WINDOW_SIZE { " (?)" } else { "" },
+                elapsed.as_micros(),
+                &self.message_history,
             );
 
             self.current = 0;
+
+            // Bytes metrics
+            self.byte_history.write(self.bytes);
+
+            log_metric(
+                "bytes processed",
+                self.bytes,
+                elapsed.as_micros(),
+                &self.byte_history,
+            );
+
+            self.bytes = 0;
             self.start_time = now;
+
+            // System metrics
+            log_sys_metrics();
```

Comment:
I'm not a fan of doing this in the fast paths of pub and sub.

Reply:
good point, is there a better control plane location that these could be emitted from?

Reply:
Well, you can just `std::thread::spawn()` a thread for it that does `loop { log_system_metrics(); sleep(1s); }`.

(Could make it fancier to check for a shutdown signal so that the thread doesn't outlive `main()` and crash, but it's just an example so who cares.)

Reply:
Sounds good to me, wasn't sure if you had a larger design in plan for example, so didn't want to just put it wherever :P